### PR TITLE
Adding Mixpanel logging to ingest pipeline expression matrix errors (SCP-4037)

### PR DESCRIPTION
**Pull request by jlchang**
Adding Mixpanel logging to ingest pipeline expression matrix errors. This PR addresses logging of errors from the GeneExpression Class. 

This PR generates a "file-validation" event to mixpanel logging upon completion of an ingest pipeline expression matrix ingest.

Bugfix - with this PR file-validation events should now correctly discriminate between triggers "sync" and "upload"

Logging refinement - with this PR file-validation events will no longer report "errors" or "warnings" (but will continue to report "numErrors" and "numWarnings", where applicable) - this change avoids the loss of logging when the reported errors/warnings are too long and would trigger a Bard error instead of logging to Mixpanel

set your local Portal instance to use docker image:
gcr.io/broad-singlecellportal-staging/scp-ingest-jlc_matrix_mixpanel:58a9a62
	
Test 1: show successful ingest of a sparse matrix using upload wizard

a. upload the following sparse matrix files from the scp-ingest-pipeline repo
tests/data/mtx/matrix.mtx
tests/data/mtx/barcodes.tsv
tests/data/mtx/genes.tsv

b. confirm that the mixpanel file-validation event (in Mixpanel Dev) with 
fileType: MM Coordinate Matrix
logger: ingest-pipeline 
trigger: upload
status: success

Test 2: show logging of failed ingest of a dense matrix using upload wizard

a. gsutil the following dense matrix files from the scp-ingest-pipeline repo
tests/data/expression_matrix_non_unique.txt

b. sync the file as a raw count matrix (or processed matrix) should fail either way

c. confirm that the mixpanel file-validation event (in Mixpanel Dev) with 
confirm that field 'errors' is not reported
errorTypes: ["content:duplicate:cells-within-file"]
numErrors: 1
numErrortypes: 1
fileType: Expression Matrix
logger: ingest-pipeline 
trigger: sync
status: failure

This PR satisfies SCP-4037.

### CODE REVIEW

File: ingest/expression_files/dense_ingestor.py [hunk 1]
```diff
@@ -188,15 +194,27 @@ def filter_expression_scores(scores: List, cells: List):
...
+                    "error", "content:type:not-numeric", msg
+                )
                 raise ValueError("Score '{expression_score}' is not valid")
         return valid_expression_scores, associated_cells
 
     @staticmethod
     def check_unique_header(header: List):
         """Confirms header has no duplicate values"""
         if len(set(header)) != len(header):
+            msg = "Duplicate header values are not allowed"
```
Comment (eweitz) [inline ingest/expression_files/dense_ingestor.py:214]: Possible separate task: tell the users [which values are duplicates](https://stackoverflow.com/a/9835819) (perhaps up to 10 duplicates).

File: ingest/expression_files/dense_ingestor.py [hunk 2]
```diff
@@ -224,7 +246,9 @@ def check_gene_keyword(header: List, row: List):
             return True
         if DenseIngestor.is_r_formatted_file(header, row)[0]:
             return True
-        raise ValueError("Required 'GENE' header is not present")
+        msg = "Required 'GENE' header is not present"
```
Comment (eweitz) [inline ingest/expression_files/dense_ingestor.py:249]: Double-quotes for user-facing text:

```suggestion
        msg = "Required \"GENE\" header is not present"
```

File: ingest/expression_files/mtx.py [hunk 3]
```diff
@@ -201,14 +211,20 @@ def get_data_start_line_number(file_handler: IO) -> int:
                     # First line w/o '%' is mtx dimension. So skip this line (+1)
                     return count + 1
                 except ValueError:
-                    raise ValueError(
-                        "Only header, comment lines starting with '%', and numeric data allowed in MTX file."
+                    msg = "Only header, comment lines starting with '%', and numeric data allowed in MTX file."
```
Comment (eweitz) [inline ingest/expression_files/mtx.py:214]:
```suggestion
                    msg = "Only header, comment lines starting with \"%\", and numeric data allowed in MTX file."
```

File: ingest/expression_files/mtx.py [hunk 4]
```diff
@@ -221,7 +237,9 @@ def get_mtx_dimensions(file_handler) -> List:
                     return dimensions
                 except Exception as e:
                     raise e
-        raise ValueError("MTX file did not contain data")
+        msg = "MTX file did not contain data"
```
Comment (eweitz) [inline ingest/expression_files/mtx.py:240]: Right?

```suggestion
        msg = "MTX file did not contain expression data"
```

File: ingest/ingest_pipeline.py [hunk 5]
```diff
@@ -8,18 +8,7 @@
...
-    export MONGODB_USERNAME='single_cell'
-    export DATABASE_NAME='single_cell_portal_development'
-leverage Vault to securely set up the following environment variable(s):
-    MONGODB_PASSWORD
-    DATABASE_HOST
-Bypass MongoDB check for ingest file info in MongoDB and bypass writing results to MongoDB
-    export BYPASS_MONGO_WRITES='yes'
-(optional) To have validation events report to Mixpanel, set:
-    export BARD_HOST_URL="https://terra-bard-dev.appspot.com"
+DEVELOPER SETUP (see README.md/Install and ../scripts/setup_mongo_dev.sh)
```
Comment (eweitz) [inline ingest/ingest_pipeline.py:11]: More copy-pasteable into browser address bar:

```suggestion
DEVELOPER SETUP (see README.md#Install and ../scripts/setup_mongo_dev.sh)
```

**Review by eweitz — APPROVED** [review]
Code looks good!  Thanks for following through on the deferred suggestions from #228.

I suggest a handful of refinements, none blocking.

File: README.md [hunk 6]
```diff
@@ -59,6 +60,20 @@ vault read secret/kdux/scp/development/$BROAD_USER/mongo/hostname
...
 export DATABASE_HOST="<ip from Vault (omit brackets)>"
 ```
+For testing/development using ingest_pipeline.py on the command line, annotation file input validation and MongoDB writes can be bypassed if you set:
+(Note: this bypass does not currently apply to expression matrix ingest).
+```
+export BYPASS_MONGO_WRITES='yes'
+```
+
+
+If you are developing updates for MixPanel logging, set the Bard Host URL:
```
[the remark this answers is not in the document]
Comment (jlchang) [reply README.md]: `d18014a`

File: ingest/expression_files/dense_ingestor.py [hunk 7]
```diff
@@ -188,15 +194,27 @@ def filter_expression_scores(scores: List, cells: List):
...
                         associated_cells.append(cells[idx])
             except ValueError as err:
-                raise ValueError(f"Expected numeric expression score - {err}")
+                msg = f"Expected numeric expression score - {err}"
+                GeneExpression.log_for_mixpanel(
+                    "error", "content:type:not-numeric", msg
+                )
+                raise ValueError(msg)
             except Exception:
+                msg = "Score '{expression_score}' is not valid"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/expression_files/dense_ingestor.py:203]: `d18014a`

File: ingest/expression_files/dense_ingestor.py (hunk 1 above)
Comment (jlchang) [reply ingest/expression_files/dense_ingestor.py]: Good point. More helpful to provide some examples of the duplication. I've added three example duplicates to the error message.
`d18014a`

File: ingest/expression_files/dense_ingestor.py [hunk 8]
```diff
@@ -205,9 +223,13 @@ def check_header_valid_values(header: List[str]):
         """Validates there are no empty header values"""
         for value in header:
             if value == "" or value.isspace():
-                raise ValueError("Header values cannot be blank")
+                msg = "Header values cannot be blank"
+                GeneExpression.log_for_mixpanel("error", "format:cap:no-empty", msg)
+                raise ValueError(msg)
             if value.lower() == "nan":
-                raise ValueError(f"{value} is not allowed as a header value")
+                msg = f"{value} is not allowed as a header value"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/expression_files/dense_ingestor.py]: `d18014a`

File: ingest/expression_files/dense_ingestor.py (hunk 2 above)
Comment (jlchang) [reply ingest/expression_files/dense_ingestor.py]: `d18014a`

File: ingest/expression_files/mtx.py (hunk 3 above)
Comment (jlchang) [reply ingest/expression_files/mtx.py]: `d18014a`

File: ingest/expression_files/mtx.py (hunk 4 above)
Comment (jlchang) [reply ingest/expression_files/mtx.py]: Correct! Thanks for improving the error message to be more precise.`d18014a`

File: ingest/ingest_pipeline.py (hunk 5 above)
Comment (jlchang) [reply ingest/ingest_pipeline.py]: Had not realized that was possible, thanks! `d18014a`

**Review by bistline — APPROVED** [review]
Looks good - some minor nits about message formatting.  Of more concern is that I'm not seeing the events in Mixpanel dev, but that could just be latency.  Apparently this happens sometimes (there haven't been any new events in dev or prod MP for ~30m)

File: ingest/cell_metadata.py [hunk 9]
```diff
@@ -109,7 +109,7 @@ def validate_header_for_coordinate_values(self):
         if valid:
             return True
         else:
-            msg = "Header names can not be coordinate values x, y, or z (case insensitive)"
+            msg = "Header names can not be coordinate values \"X\", \"Y\", or \"Z\" (case insensitive)"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/cell_metadata.py]: agreed, not sure why I didn't think to change the outermost quotes... `334b09a`

File: ingest/clusters.py [hunk 10]
```diff
@@ -111,9 +111,7 @@ def validate_header_for_coordinate_values(self):
         lower_cased_headers = [header.lower() for header in self.headers]
         for coordinate in ("x", "y"):
             if coordinate not in lower_cased_headers:
-                msg = (
-                    "Header must have coordinate values 'x' and 'y' (case insensitive)"
-                )
+                msg = "Header must have coordinate values \"X\" and \"Y\" (case insensitive)"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/clusters.py]: `334b09a`

File: ingest/expression_files/dense_ingestor.py [hunk 11]
```diff
@@ -224,7 +249,9 @@ def check_gene_keyword(header: List, row: List):
             return True
         if DenseIngestor.is_r_formatted_file(header, row)[0]:
             return True
-        raise ValueError("Required 'GENE' header is not present")
+        msg = "Required \"GENE\" header is not present"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/expression_files/dense_ingestor.py]: `334b09a`

File: ingest/expression_files/mtx.py [hunk 12]
```diff
@@ -201,14 +211,20 @@ def get_data_start_line_number(file_handler: IO) -> int:
                     # First line w/o '%' is mtx dimension. So skip this line (+1)
                     return count + 1
                 except ValueError:
-                    raise ValueError(
-                        "Only header, comment lines starting with '%', and numeric data allowed in MTX file."
+                    msg = "Only header, comment lines starting with \"%\", and numeric data allowed in MTX file."
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/expression_files/mtx.py]: `334b09a`

File: ingest/config.py [hunk 13]
```diff
@@ -92,7 +120,7 @@ def study(self, study_id: str):
         try:
             study_id = ObjectId(study_id)
         except Exception:
-            raise ValueError("Must pass in valid object ID for study ID")
+            raise ValueError("Must pass in valid object ID for study ID.")
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/config.py:123]: Agreed, I went through and added a period where appropriate (it was a pain to chase down all the broken tests but we should now be more consistent) `334b09a`

File: tests/test_dense.py [hunk 14]
```diff
@@ -230,14 +230,16 @@ def test_check_valid(self, mock_check_unique_cells):
             DenseIngestor.check_valid(
                 ["GENE", "foo", "foo"], ["gene1", "0", "1"], query_params
             )
-        expected_msg = "Duplicate header values are not allowed"
+        expected_msg = (
+            "Duplicate header values are not allowed. Duplicates include: \"foo\""
```
[the remark this answers is not in the document]
Comment (jlchang) [reply tests/test_dense.py]: `334b09a`

File: tests/test_dense.py [hunk 15]
```diff
@@ -230,14 +230,16 @@ def test_check_valid(self, mock_check_unique_cells):
...
+        )
         self.assertEqual(expected_msg, str(cm.exception))
 
         # Should concatenate the two value errors
         mock_check_unique_cells.return_value = True
         with self.assertRaises(ValueError) as cm:
             DenseIngestor.check_valid(["foo", "nan"], ["foo2", "foo3"], query_params)
-        expected_msg = "Required 'GENE' header is not present; nan is not allowed as a header value"
+        expected_msg = "Required \"GENE\" header is not present; \"nan\" is not allowed as a header value"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply tests/test_dense.py]: `334b09a`

**Pull request merged**